File: 039_Ransom Note.py
"""
Ransom Note

Given two strings ransomNote and magazine, return true if ransomNote can be constructed
from the letters in magazine. Each letter in magazine can only be used once.
"""

def can_construct(ransom_note: str, magazine: str) -> bool:
    from collections import Counter
    need = Counter(ransom_note)
    avail = Counter(magazine)
    for k, v in need.items():
        if avail[k] < v: return False
    return True


if __name__ == '__main__':
    print('Ransom 1:', can_construct('a', 'b'))
    print('Ransom 2:', can_construct('aa', 'ab'))
    print('Ransom 3:', can_construct('aa', 'aab'))

# Time Complexity: O(n + m)
# Space Complexity: O(1) (alphabet constant)

File: 039_Ransom Note.cpp
/*
Ransom Note

Given two strings ransomNote and magazine, return true if ransomNote can be constructed
from the letters in magazine. Each letter in magazine can only be used once.
*/

#include <bits/stdc++.h>
using namespace std;

bool canConstruct(string ransomNote, string magazine) {
    vector<int> cnt(26,0);
    for (char c: magazine) cnt[c - 'a']++;
    for (char c: ransomNote) {
        if (--cnt[c - 'a'] < 0) return false;
    }
    return true;
}

int main() {
    cout << "Ransom 1: " << (canConstruct("a","b")?"true":"false") << endl; // false
    cout << "Ransom 2: " << (canConstruct("aa","ab")?"true":"false") << endl; // false
    cout << "Ransom 3: " << (canConstruct("aa","aab")?"true":"false") << endl; // true
    return 0;
}

// Time Complexity: O(n + m)
// Space Complexity: O(1)
